File: tree.js
// Family Tree JavaScript Implementation
class FamilyTree {
    constructor() {
        this.members = new Map();
        this.root = null;
        this.currentEditingMember = null;
        this.svg = null;
        this.g = null;
        this.zoom = null;
        this.width = 0;
        this.height = 0;
        
        this.init();
    }

    init() {
        // Initialize Firebase auth check
        firebase.auth().onAuthStateChanged((user) => {
            if (!user) {
                window.location.href = 'user-login.html';
                return;
            }
            this.loadFamilyTree();
        });

        this.setupEventListeners();
        this.initializeSVG();
        this.loadSampleData(); // For demo purposes
    }

    setupEventListeners() {
        // Modal controls
        document.getElementById('add-member-btn').addEventListener('click', () => this.showAddMemberModal());
        document.getElementById('add-first-member-btn').addEventListener('click', () => this.showAddMemberModal());
        document.getElementById('close-modal').addEventListener('click', () => this.hideModal());
        document.getElementById('cancel-btn').addEventListener('click', () => this.hideModal());
        
        // Form submission
        document.getElementById('member-form').addEventListener('submit', (e) => this.handleFormSubmit(e));
        
        // Tree controls
        document.getElementById('center-tree-btn').addEventListener('click', () => this.centerTree());
        document.getElementById('expand-all-btn').addEventListener('click', () => this.expandAll());
        document.getElementById('collapse-all-btn').addEventListener('click', () => this.collapseAll());
        
        // Context menu
        document.addEventListener('click', () => this.hideContextMenu());
        document.getElementById('edit-member').addEventListener('click', () => this.editSelectedMember());
        document.getElementById('add-child').addEventListener('click', () => this.addChildToSelected());
        document.getElementById('delete-member').addEventListener('click', () => this.deleteSelectedMember());
        
        // Close modal on outside click
        document.getElementById('member-modal').addEventListener('click', (e) => {
            if (e.target.id === 'member-modal') {
                this.hideModal();
            }
        });

        // Logout functionality
        const logoutBtn = document.getElementById('logout-btn');
        if (logoutBtn) {
            logoutBtn.addEventListener('click', (e) => {
                e.preventDefault();
                firebase.auth().signOut().then(() => {
                    window.location.href = 'user-login.html';
                }).catch(error => {
                    console.error('Logout error:', error);
                });
            });
        }
    }

    initializeSVG() {
        const container = document.getElementById('tree-svg-container');
        this.width = container.clientWidth;
        this.height = container.clientHeight;

        this.svg = d3.select('#family-tree-svg')
            .attr('width', this.width)
            .attr('height', this.height);

        // Create zoom behavior
        this.zoom = d3.zoom()
            .scaleExtent([0.1, 3])
            .on('zoom', (event) => {
                this.g.attr('transform', event.transform);
            });

        this.svg.call(this.zoom);

        // Create main group for tree content
        this.g = this.svg.append('g')
            .attr('class', 'tree-content');

        // Add zoom controls
        this.addZoomControls();
    }

    addZoomControls() {
        const controls = d3.select('#tree-svg-container')
            .append('div')
            .attr('class', 'zoom-controls');

        controls.append('button')
            .attr('class', 'zoom-btn')
            .html('<i class="fas fa-plus"></i>')
            .on('click', () => {
                this.svg.transition().duration(300).call(
                    this.zoom.scaleBy, 1.5
                );
            });

        controls.append('button')
            .attr('class', 'zoom-btn')
            .html('<i class="fas fa-minus"></i>')
            .on('click', () => {
                this.svg.transition().duration(300).call(
                    this.zoom.scaleBy, 1 / 1.5
                );
            });
    }

    loadSampleData() {
        // Add sample family data for demonstration
        this.addMember('John Smith', null, 'Root', '1950-01-15', 'Family patriarch');
        this.addMember('Mary Smith', 'John Smith', 'Spouse', '1952-03-20', 'Loving wife and mother');
        this.addMember('Alice Johnson', 'John Smith', 'Child', '1975-07-10', 'Eldest daughter');
        this.addMember('Bob Smith', 'John Smith', 'Child', '1978-11-05', 'Son');
        this.addMember('Charlie Johnson', 'Alice Johnson', 'Child', '2000-04-12', 'Grandson');
        this.addMember('Diana Johnson', 'Alice Johnson', 'Child', '2003-09-18', 'Granddaughter');
        
        this.updateVisualization();
        this.updateStats();
    }

    addMember(name, parentName, relationship, birthDate = '', notes = '') {
        if (this.members.has(name)) {
            this.showMessage('Member already exists!', 'error');
            return false;
        }

        const member = {
            name: name,
            parent: parentName,
            relationship: relationship,
            birthDate: birthDate,
            notes: notes,
            children: [],
            x: 0,
            y: 0,
            collapsed: false
        };

        this.members.set(name, member);

        // Set as root if no parent or first member
        if (!parentName || this.members.size === 1) {
            this.root = member;
            member.relationship = 'Root';
        } else {
            // Add to parent's children
            const parent = this.members.get(parentName);
            if (parent) {
                parent.children.push(member);
            }
        }

        return true;
    }

    editMember(oldName, newName, parentName, relationship, birthDate, notes) {
        const member = this.members.get(oldName);
        if (!member) {
            this.showMessage('Member not found!', 'error');
            return false;
        }

        // If name is changing, update the map
        if (oldName !== newName) {
            if (this.members.has(newName)) {
                this.showMessage('A member with this name already exists!', 'error');
                return false;
            }
            
            this.members.delete(oldName);
            member.name = newName;
            this.members.set(newName, member);
            
            // Update children references
            this.members.forEach(m => {
                if (m.parent === oldName) {
                    m.parent = newName;
                }
            });
        }

        // Update other properties
        member.relationship = relationship;
        member.birthDate = birthDate;
        member.notes = notes;

        // Handle parent change
        if (member.parent !== parentName) {
            // Remove from old parent
            if (member.parent) {
                const oldParent = this.members.get(member.parent);
                if (oldParent) {
                    oldParent.children = oldParent.children.filter(child => child !== member);
                }
            }

            // Add to new parent
            member.parent = parentName;
            if (parentName) {
                const newParent = this.members.get(parentName);
                if (newParent) {
                    newParent.children.push(member);
                }
            } else {
                this.root = member;
                member.relationship = 'Root';
            }
        }

        return true;
    }

    deleteMember(name) {
        const member = this.members.get(name);
        if (!member) {
            this.showMessage('Member not found!', 'error');
            return false;
        }

        // Recursively delete all descendants
        const deleteRecursive = (m) => {
            m.children.forEach(child => deleteRecursive(child));
            this.members.delete(m.name);
        };

        // Remove from parent's children
        if (member.parent) {
            const parent = this.members.get(member.parent);
            if (parent) {
                parent.children = parent.children.filter(child => child !== member);
            }
        }

        // If deleting root, clear tree
        if (member === this.root) {
            this.members.clear();
            this.root = null;
        } else {
            deleteRecursive(member);
        }

        return true;
    }

    updateVisualization() {
        if (!this.root) {
            this.showEmptyState();
            return;
        }

        this.hideEmptyState();

        // Create tree layout
        const treeLayout = d3.tree()
            .size([this.height - 100, this.width - 200])
            .separation((a, b) => a.parent === b.parent ? 1 : 2);

        // Convert our data structure to d3 hierarchy
        const hierarchy = this.createHierarchy(this.root);
        const treeData = treeLayout(hierarchy);

        // Get nodes and links
        const nodes = treeData.descendants();
        const links = treeData.descendants().slice(1);

        // Update links
        const link = this.g.selectAll('.tree-link')
            .data(links);

        link.enter()
            .append('path')
            .attr('class', 'tree-link')
            .merge(link)
            .transition()
            .duration(500)
            .attr('d', d => {
                return `M${d.y},${d.x}
                        C${(d.y + d.parent.y) / 2},${d.x}
                         ${(d.y + d.parent.y) / 2},${d.parent.x}
                         ${d.parent.y},${d.parent.x}`;
            });

        link.exit().remove();

        // Update nodes
        const node = this.g.selectAll('.tree-node')
            .data(nodes);

        const nodeEnter = node.enter()
            .append('g')
            .attr('class', 'tree-node')
            .attr('transform', d => `translate(${d.y},${d.x})`)
            .on('click', (event, d) => this.handleNodeClick(event, d))
            .on('contextmenu', (event, d) => this.handleNodeRightClick(event, d));

        // Add circles
        nodeEnter.append('circle')
            .attr('class', d => `node-circle ${d.data.relationship === 'Root' ? 'root' : ''}`)
            .attr('r', 25);

        // Add names
        nodeEnter.append('text')
            .attr('class', d => `node-text ${d.data.relationship === 'Root' ? 'root' : ''}`)
            .attr('dy', '-0.3em')
            .text(d => d.data.name);

        // Add relationships
        nodeEnter.append('text')
            .attr('class', 'node-relationship')
            .attr('dy', '1.2em')
            .text(d => d.data.relationship !== 'Root' ? d.data.relationship : '');

        // Update existing nodes
        node.merge(nodeEnter)
            .transition()
            .duration(500)
            .attr('transform', d => `translate(${d.y},${d.x})`);

        node.exit().remove();

        // Center the tree
        this.centerTree();
    }

    createHierarchy(rootMember) {
        const createNode = (member) => {
            const node = {
                name: member.name,
                relationship: member.relationship,
                birthDate: member.birthDate,
                notes: member.notes,
                children: member.children.map(child => createNode(child))
            };
            return node;
        };

        return d3.hierarchy(createNode(rootMember));
    }

    handleNodeClick(event, d) {
        event.stopPropagation();
        // Double click to edit
        if (event.detail === 2) {
            this.showEditMemberModal(d.data.name);
        }
    }

    handleNodeRightClick(event, d) {
        event.preventDefault();
        event.stopPropagation();
        this.showContextMenu(event, d.data.name);
    }

    showContextMenu(event, memberName) {
        this.selectedMember = memberName;
        const contextMenu = document.getElementById('context-menu');
        contextMenu.style.display = 'block';
        contextMenu.style.left = event.pageX + 'px';
        contextMenu.style.top = event.pageY + 'px';
    }

    hideContextMenu() {
        document.getElementById('context-menu').style.display = 'none';
    }

    showAddMemberModal(parentName = null) {
        this.currentEditingMember = null;
        document.getElementById('modal-title').textContent = 'Add Family Member';
        document.getElementById('member-form').reset();
        
        // Populate parent dropdown
        this.populateParentDropdown();
        
        if (parentName) {
            document.getElementById('member-parent').value = parentName;
        }
        
        document.getElementById('member-modal').style.display = 'block';
    }

    showEditMemberModal(memberName) {
        const member = this.members.get(memberName);
        if (!member) return;

        this.currentEditingMember = memberName;
        document.getElementById('modal-title').textContent = 'Edit Family Member';
        
        // Populate form
        document.getElementById('member-name').value = member.name;
        document.getElementById('member-relationship').value = member.relationship;
        document.getElementById('member-birthdate').value = member.birthDate;
        document.getElementById('member-notes').value = member.notes;
        
        this.populateParentDropdown(memberName);
        document.getElementById('member-parent').value = member.parent || '';
        
        document.getElementById('member-modal').style.display = 'block';
    }

    populateParentDropdown(excludeName = null) {
        const select = document.getElementById('member-parent');
        select.innerHTML = '<option value="">Select Parent (Optional)</option>';
        
        this.members.forEach((member, name) => {
            if (name !== excludeName) {
                const option = document.createElement('option');
                option.value = name;
                option.textContent = name;
                select.appendChild(option);
            }
        });
    }

    hideModal() {
        document.getElementById('member-modal').style.display = 'none';
        this.currentEditingMember = null;
    }

    handleFormSubmit(event) {
        event.preventDefault();
        
        const formData = new FormData(event.target);
        const name = formData.get('name').trim();
        const parent = formData.get('parent') || null;
        const relationship = formData.get('relationship');
        const birthDate = formData.get('birthdate');
        const notes = formData.get('notes');

        if (!name) {
            this.showMessage('Name is required!', 'error');
            return;
        }

        let success = false;
        
        if (this.currentEditingMember) {
            // Edit existing member
            success = this.editMember(this.currentEditingMember, name, parent, relationship, birthDate, notes);
            if (success) {
                this.showMessage('Member updated successfully!', 'success');
            }
        } else {
            // Add new member
            success = this.addMember(name, parent, relationship, birthDate, notes);
            if (success) {
                this.showMessage('Member added successfully!', 'success');
            }
        }

        if (success) {
            this.hideModal();
            this.updateVisualization();
            this.updateStats();
            this.saveFamilyTree();
        }
    }

    editSelectedMember() {
        if (this.selectedMember) {
            this.showEditMemberModal(this.selectedMember);
            this.hideContextMenu();
        }
    }

    addChildToSelected() {
        if (this.selectedMember) {
            this.showAddMemberModal(this.selectedMember);
            this.hideContextMenu();
        }
    }

    deleteSelectedMember() {
        if (this.selectedMember) {
            const member = this.members.get(this.selectedMember);
            const childCount = member ? member.children.length : 0;
            
            let message = `Are you sure you want to delete ${this.selectedMember}?`;
            if (childCount > 0) {
                message += `\n\nThis will also delete ${childCount} descendant(s).`;
            }
            
            if (confirm(message)) {
                if (this.deleteMember(this.selectedMember)) {
                    this.showMessage('Member deleted successfully!', 'success');
                    this.updateVisualization();
                    this.updateStats();
                    this.saveFamilyTree();
                }
            }
            this.hideContextMenu();
        }
    }

    centerTree() {
        if (!this.root) return;
        
        const bounds = this.g.node().getBBox();
        const fullWidth = this.width;
        const fullHeight = this.height;
        const width = bounds.width;
        const height = bounds.height;
        const midX = bounds.x + width / 2;
        const midY = bounds.y + height / 2;
        
        if (width === 0 || height === 0) return;
        
        const scale = 0.8 / Math.max(width / fullWidth, height / fullHeight);
        const translate = [fullWidth / 2 - scale * midX, fullHeight / 2 - scale * midY];
        
        this.svg.transition()
            .duration(750)
            .call(this.zoom.transform, d3.zoomIdentity.translate(translate[0], translate[1]).scale(scale));
    }

    expandAll() {
        this.members.forEach(member => {
            member.collapsed = false;
        });
        this.updateVisualization();
    }

    collapseAll() {
        this.members.forEach(member => {
            if (member !== this.root) {
                member.collapsed = true;
            }
        });
        this.updateVisualization();
    }

    showEmptyState() {
        document.getElementById('empty-state').style.display = 'flex';
    }

    hideEmptyState() {
        document.getElementById('empty-state').style.display = 'none';
    }

    updateStats() {
        const totalMembers = this.members.size;
        const generations = this.calculateGenerations();
        const lastUpdated = new Date().toLocaleDateString();

        document.getElementById('total-members').textContent = totalMembers;
        document.getElementById('total-generations').textContent = generations;
        document.getElementById('last-updated').textContent = lastUpdated;
    }

    calculateGenerations() {
        if (!this.root) return 0;
        
        const getDepth = (member, depth = 1) => {
            if (member.children.length === 0) return depth;
            return Math.max(...member.children.map(child => getDepth(child, depth + 1)));
        };
        
        return getDepth(this.root);
    }

    showMessage(text, type = 'success') {
        const message = document.createElement('div');
        message.className = `message ${type}`;
        message.textContent = text;
        document.body.appendChild(message);
        
        setTimeout(() => {
            message.remove();
        }, 3000);
    }

    saveFamilyTree() {
        // In a real application, save to Firebase
        const user = firebase.auth().currentUser;
        if (user) {
            const treeData = {
                members: Array.from(this.members.entries()),
                rootName: this.root ? this.root.name : null,
                lastUpdated: new Date().toISOString()
            };
            
            // Save to localStorage for demo
            localStorage.setItem(`familyTree_${user.uid}`, JSON.stringify(treeData));
            
            // In production, save to Firestore:
            // firebase.firestore().collection('familyTrees').doc(user.uid).set(treeData);
        }
    }

    loadFamilyTree() {
        const user = firebase.auth().currentUser;
        if (user) {
            // Load from localStorage for demo
            const savedData = localStorage.getItem(`familyTree_${user.uid}`);
            if (savedData) {
                try {
                    const treeData = JSON.parse(savedData);
                    this.members = new Map(treeData.members);
                    this.root = treeData.rootName ? this.members.get(treeData.rootName) : null;
                    
                    // Rebuild parent-child relationships
                    this.members.forEach(member => {
                        member.children = [];
                    });
                    
                    this.members.forEach(member => {
                        if (member.parent) {
                            const parent = this.members.get(member.parent);
                            if (parent) {
                                parent.children.push(member);
                            }
                        }
                    });
                    
                    this.updateVisualization();
                    this.updateStats();
                    return;
                } catch (error) {
                    console.error('Error loading family tree:', error);
                }
            }
        }
        
        // If no saved data, load sample data
        this.loadSampleData();
    }
}

// Initialize the family tree when the page loads
document.addEventListener('DOMContentLoaded', () => {
    new FamilyTree();
});

File: Tree.java
import java.util.*;

/**
 * Family Tree Implementation
 * Represents a family tree where each node contains a person's information
 * and can have multiple children (descendants).
 */
public class Tree {
    private TreeNode root;
    private Map<String, TreeNode> nodeMap; // For quick lookup by name
    private int totalMembers;

    /**
     * Inner class representing a person in the family tree
     */
    public static class TreeNode {
        private String name;
        private String birthDate;
        private String relationship;
        private TreeNode parent;
        private List<TreeNode> children;
        private Map<String, String> additionalInfo;

        public TreeNode(String name) {
            this.name = name;
            this.children = new ArrayList<>();
            this.additionalInfo = new HashMap<>();
        }

        public TreeNode(String name, String birthDate, String relationship) {
            this(name);
            this.birthDate = birthDate;
            this.relationship = relationship;
        }

        // Getters and Setters
        public String getName() { return name; }
        public void setName(String name) { this.name = name; }
        
        public String getBirthDate() { return birthDate; }
        public void setBirthDate(String birthDate) { this.birthDate = birthDate; }
        
        public String getRelationship() { return relationship; }
        public void setRelationship(String relationship) { this.relationship = relationship; }
        
        public TreeNode getParent() { return parent; }
        public void setParent(TreeNode parent) { this.parent = parent; }
        
        public List<TreeNode> getChildren() { return children; }
        
        public Map<String, String> getAdditionalInfo() { return additionalInfo; }
        
        public void addChild(TreeNode child) {
            if (child != null && !children.contains(child)) {
                children.add(child);
                child.setParent(this);
            }
        }
        
        public boolean removeChild(TreeNode child) {
            if (children.remove(child)) {
                child.setParent(null);
                return true;
            }
            return false;
        }
        
        public void addInfo(String key, String value) {
            additionalInfo.put(key, value);
        }
        
        public String getInfo(String key) {
            return additionalInfo.get(key);
        }
    }

    /**
     * Constructor - Initialize empty family tree
     */
    public Tree() {
        this.nodeMap = new HashMap<>();
        this.totalMembers = 0;
    }

    /**
     * Constructor with root person
     */
    public Tree(String rootName) {
        this();
        this.root = new TreeNode(rootName, null, "Root");
        nodeMap.put(rootName.toLowerCase(), root);
        totalMembers = 1;
    }

    /**
     * Add a new family member under a specific parent
     * @param childName Name of the new family member
     * @param parentName Name of the parent (null for root)
     * @param relationship Relationship to parent
     * @return true if successfully added, false otherwise
     */
    public boolean addNode(String childName, String parentName, String relationship) {
        if (childName == null || childName.trim().isEmpty()) {
            return false;
        }

        childName = childName.trim();
        
        // Check if person already exists
        if (nodeMap.containsKey(childName.toLowerCase())) {
            return false; // Person already exists
        }

        TreeNode newNode = new TreeNode(childName, null, relationship);

        // If no parent specified or tree is empty, make this the root
        if (parentName == null || parentName.trim().isEmpty() || root == null) {
            if (root == null) {
                root = newNode;
                newNode.setRelationship("Root");
            } else {
                return false; // Root already exists
            }
        } else {
            // Find parent node
            TreeNode parent = findNode(parentName.trim());
            if (parent == null) {
                return false; // Parent not found
            }
            parent.addChild(newNode);
        }

        nodeMap.put(childName.toLowerCase(), newNode);
        totalMembers++;
        return true;
    }

    /**
     * Edit an existing family member's information
     * @param oldName Current name of the person
     * @param newName New name (can be same as old)
     * @param newBirthDate New birth date
     * @param newRelationship New relationship
     * @return true if successfully edited, false otherwise
     */
    public boolean editNode(String oldName, String newName, String newBirthDate, String newRelationship) {
        if (oldName == null || oldName.trim().isEmpty()) {
            return false;
        }

        TreeNode node = findNode(oldName.trim());
        if (node == null) {
            return false; // Person not found
        }

        // If name is changing, update the map
        if (newName != null && !newName.trim().isEmpty() && !newName.trim().equals(oldName.trim())) {
            String newNameLower = newName.trim().toLowerCase();
            
            // Check if new name already exists
            if (nodeMap.containsKey(newNameLower)) {
                return false; // New name already exists
            }
            
            // Remove old mapping and add new one
            nodeMap.remove(oldName.trim().toLowerCase());
            node.setName(newName.trim());
            nodeMap.put(newNameLower, node);
        }

        // Update other information
        if (newBirthDate != null) {
            node.setBirthDate(newBirthDate.trim());
        }
        
        if (newRelationship != null && !newRelationship.trim().isEmpty()) {
            node.setRelationship(newRelationship.trim());
        }

        return true;
    }

    /**
     * Delete a family member and all their descendants
     * @param name Name of the person to delete
     * @return true if successfully deleted, false otherwise
     */
    public boolean deleteNode(String name) {
        if (name == null || name.trim().isEmpty()) {
            return false;
        }

        TreeNode nodeToDelete = findNode(name.trim());
        if (nodeToDelete == null) {
            return false; // Person not found
        }

        // Count nodes to be deleted (including descendants)
        int deletedCount = countDescendants(nodeToDelete) + 1;

        // If deleting root, clear entire tree
        if (nodeToDelete == root) {
            clear();
            return true;
        }

        // Remove from parent's children list
        TreeNode parent = nodeToDelete.getParent();
        if (parent != null) {
            parent.removeChild(nodeToDelete);
        }

        // Remove from map (including all descendants)
        removeFromMap(nodeToDelete);
        
        totalMembers -= deletedCount;
        return true;
    }

    /**
     * Find a node by name
     * @param name Name to search for
     * @return TreeNode if found, null otherwise
     */
    public TreeNode findNode(String name) {
        if (name == null || name.trim().isEmpty()) {
            return null;
        }
        return nodeMap.get(name.trim().toLowerCase());
    }

    /**
     * Get the tree structure as a formatted string for display
     * @return String representation of the tree
     */
    public String displayTree() {
        if (root == null) {
            return "Family tree is empty.";
        }
        
        StringBuilder sb = new StringBuilder();
        sb.append("Family Tree Structure:\n");
        sb.append("======================\n");
        displayTreeHelper(root, "", true, sb);
        sb.append("\nTotal family members: ").append(totalMembers);
        return sb.toString();
    }

    /**
     * Get tree structure as JSON-like format for web display
     * @return JSON string representation
     */
    public String getTreeAsJSON() {
        if (root == null) {
            return "{}";
        }
        
        StringBuilder sb = new StringBuilder();
        buildJSONHelper(root, sb);
        return sb.toString();
    }

    /**
     * Get all family members as a list
     * @return List of all family member names
     */
    public List<String> getAllMembers() {
        List<String> members = new ArrayList<>();
        if (root != null) {
            collectAllMembers(root, members);
        }
        return members;
    }

    /**
     * Get children of a specific person
     * @param parentName Name of the parent
     * @return List of children names
     */
    public List<String> getChildren(String parentName) {
        TreeNode parent = findNode(parentName);
        List<String> childrenNames = new ArrayList<>();
        
        if (parent != null) {
            for (TreeNode child : parent.getChildren()) {
                childrenNames.add(child.getName());
            }
        }
        
        return childrenNames;
    }

    /**
     * Get parent of a specific person
     * @param childName Name of the child
     * @return Parent name or null if no parent
     */
    public String getParent(String childName) {
        TreeNode child = findNode(childName);
        if (child != null && child.getParent() != null) {
            return child.getParent().getName();
        }
        return null;
    }

    /**
     * Clear the entire tree
     */
    public void clear() {
        root = null;
        nodeMap.clear();
        totalMembers = 0;
    }

    /**
     * Get total number of family members
     * @return Total count
     */
    public int getTotalMembers() {
        return totalMembers;
    }

    /**
     * Get root node
     * @return Root TreeNode
     */
    public TreeNode getRoot() {
        return root;
    }

    // Helper methods

    private void displayTreeHelper(TreeNode node, String prefix, boolean isLast, StringBuilder sb) {
        if (node == null) return;

        sb.append(prefix);
        sb.append(isLast ? "└── " : "├── ");
        sb.append(node.getName());
        
        if (node.getBirthDate() != null && !node.getBirthDate().isEmpty()) {
            sb.append(" (").append(node.getBirthDate()).append(")");
        }
        
        if (node.getRelationship() != null && !node.getRelationship().isEmpty() && !node.getRelationship().equals("Root")) {
            sb.append(" [").append(node.getRelationship()).append("]");
        }
        
        sb.append("\n");

        List<TreeNode> children = node.getChildren();
        for (int i = 0; i < children.size(); i++) {
            boolean isLastChild = (i == children.size() - 1);
            String newPrefix = prefix + (isLast ? "    " : "│   ");
            displayTreeHelper(children.get(i), newPrefix, isLastChild, sb);
        }
    }

    private void buildJSONHelper(TreeNode node, StringBuilder sb) {
        sb.append("{");
        sb.append("\"name\":\"").append(escapeJSON(node.getName())).append("\",");
        sb.append("\"relationship\":\"").append(escapeJSON(node.getRelationship() != null ? node.getRelationship() : "")).append("\",");
        sb.append("\"birthDate\":\"").append(escapeJSON(node.getBirthDate() != null ? node.getBirthDate() : "")).append("\",");
        sb.append("\"children\":[");
        
        List<TreeNode> children = node.getChildren();
        for (int i = 0; i < children.size(); i++) {
            if (i > 0) sb.append(",");
            buildJSONHelper(children.get(i), sb);
        }
        
        sb.append("]}");
    }

    private String escapeJSON(String str) {
        if (str == null) return "";
        return str.replace("\"", "\\\"").replace("\n", "\\n").replace("\r", "\\r");
    }

    private void collectAllMembers(TreeNode node, List<String> members) {
        if (node == null) return;
        
        members.add(node.getName());
        for (TreeNode child : node.getChildren()) {
            collectAllMembers(child, members);
        }
    }

    private int countDescendants(TreeNode node) {
        if (node == null) return 0;
        
        int count = 0;
        for (TreeNode child : node.getChildren()) {
            count += 1 + countDescendants(child);
        }
        return count;
    }

    private void removeFromMap(TreeNode node) {
        if (node == null) return;
        
        nodeMap.remove(node.getName().toLowerCase());
        for (TreeNode child : node.getChildren()) {
            removeFromMap(child);
        }
    }

    // Main method for testing
    public static void main(String[] args) {
        Tree familyTree = new Tree("John Smith");
        
        // Add family members
        familyTree.addNode("Mary Smith", "John Smith", "Spouse");
        familyTree.addNode("Alice Smith", "John Smith", "Daughter");
        familyTree.addNode("Bob Smith", "John Smith", "Son");
        familyTree.addNode("Charlie Smith", "Alice Smith", "Grandson");
        
        // Display tree
        System.out.println(familyTree.displayTree());
        
        // Test editing
        familyTree.editNode("Alice Smith", "Alice Johnson", "1990-05-15", "Daughter");
        
        // Display updated tree
        System.out.println("\nAfter editing Alice:");
        System.out.println(familyTree.displayTree());
        
        // Test JSON output
        System.out.println("\nJSON representation:");
        System.out.println(familyTree.getTreeAsJSON());
    }
}